pvlib/spectrum.py

```diff
@@ -0,0 +1,318 @@
+    wavelength = _SPECTRL2_COEFFS['wavelength'][:, np.newaxis]
+    spectrum_et = _SPECTRL2_COEFFS['spectral_irradiance_et'][:, np.newaxis]
+
+    optical_thickness = (
+        aerosol_turbidity_500nm * (wavelength / 0.5)**-alpha
+    )  # Eq 2-7
+
+    # Eq 3-16
+    scattering_albedo = scattering_albedo_400nm * \
+        np.exp(-wavelength_variation_factor * np.log(wavelength / 0.4)**2)
+
+    spectrl2 = _spectrl2_transmittances(apparent_zenith, relative_airmass,
+                                        surface_pressure, precipitable_water,
+                                        ozone, optical_thickness,
+                                        scattering_albedo, dayofyear)
+    D, Tr, Ta, Tw, To, Tu, Tas, Taa = spectrl2
+
+    spectrum_et_adj = spectrum_et * D
+    # spectrum of direct irradiance, Eq 2-1
+    Id = spectrum_et_adj * Tr * Ta * Tw * To * Tu
+
+    cosZ = cosd(apparent_zenith)
+    Cs = np.where(wavelength <= 0.45, (wavelength + 0.55)**1.8, 1.0)  # Eq 3-17
+    ALG = np.log(1 - aerosol_asymmetry_factor)  # Eq 3-14
+    BFS = ALG * (0.0783 + ALG * (-0.3824 - ALG * 0.5874))  # Eq 3-13
+    AFS = ALG * (1.459 + ALG * (0.1595 + ALG * 0.4129))  # Eq 3-12
+    cosZ = cosd(apparent_zenith)
+    Fs = 1 - 0.5 * np.exp((AFS + BFS * cosZ) * cosZ)  # Eq 3-11
+    Fsp = 1 - 0.5 * np.exp((AFS + BFS / 1.8) / 1.8)  # Eq 3.15
+
+    # evaluate the "primed terms" -- transmittances evaluated at airmass=1.8
+    primes = _spectrl2_transmittances(apparent_zenith, 1.8,
+                                      surface_pressure, precipitable_water,
+                                      ozone, optical_thickness,
+                                      scattering_albedo, dayofyear)
+    _, Trp, Tap, Twp, Top, Tup, Tasp, Taap = primes
+
+    # Note: not sure what the correct form of this equation is.
+    # The first coefficient is To' in Eq 3-8 but Tu' in the code appendix.
+    # spectrl2_2.c uses Tu'.
+    sky_reflectivity = (
+        # Top * Twp * Taap * (0.5 * (1-Trp) + (1-Fsp) * Trp * (1-Tasp))
+        Tup * Twp * Taap * (0.5 * (1-Trp) + (1-Fsp) * Trp * (1-Tasp))
+    )  # Eq 3-8
+
+    # a common factor for 3-5 and 3-6
+    common_factor = spectrum_et_adj * cosZ * To * Tu * Tw * Taa
+    # Note: spectrl2_2.c differs from the report in how the Cs value is used.
+    # The two commented out lines match the report, while the following match
+    # spectrl2_2.c:
+    # Ir = common_factor * (1 - Tr**0.95) * 0.5 * Cs  # Eq 3-5
+    # Ia = common_factor * Tr**1.5 * (1 - Tas) * Fs * Cs  # Eq 3-6
+    Ir = common_factor * (1 - Tr**0.95) * 0.5  # Eq 3-5
+    Ia = common_factor * Tr**1.5 * (1 - Tas) * Fs  # Eq 3-6
+
+    rs = sky_reflectivity
+    rg = ground_albedo
+    Ig = (Id * cosZ + Ir + Ia) * rs * rg / (1 - rs * rg)  # Eq 3-7
+
+    # total scattered irradiance
+    # Note: see discussion about Cs above.
+    # Is = Ir + Ia + Ig  # Eq 3-1
+    Is = (Ir + Ia + Ig) * Cs  # Eq 3-1
+
+    # calculate spectral irradiance on a tilted surface, Eq 3-18
```

Comment:
Could transposition be a separate function?  Or could the use be expected to do it using the existing transposition functions?

Reply:
It could be a separate function, but I'm inclined to implement the complete model.

If users want to do their own transposition, is there any issue with them passing `surface_tilt=0` and transposing afterwards (other than the inefficiency of doing an unnecessary transposition in this function)?